doc/manual/contracts.md

```diff
@@ -1058,3 +1057,148 @@ is wrapping the original value with delayed checks inside**. Doing so preserves
+the previous section: this is equivalent to the contract `{ tag | 'String, value
+| String }`, and will reject the seemingly legal value `{ tag = 'Number, value =
+1+1 }`.
+
+The trick is to use a custom contract that will do more immediate checks. There
+are many different way to approach this problem, depending on the specifics of
+each situation. Here, we'll write a little `Tagged` contract wrapper that takes
+a record contract, expects `tag` to exist and be set to a fixed value, and will
+check if the provided value has a matching `tag` immediately.
+
+```nickel #repl
+> let Tagged = fun Contract =>
+  std.contract.custom (fun label =>
+    match {
+      value @ { tag, .. } =>
+        if tag == Contract.tag then
+          # from here, we just delegate the rest to Contract
+          std.contract.apply_as_custom Contract label value
+        else
+          'Error { message = "incompatible tag field" },
+      _ => 'Error { message = "missing tag field" },
+    }
+  )
+
+> let NumberOrString = std.contract.any_of [
+    Tagged { tag = 'String, value | String },
+    Tagged { tag = 'Number, value | Number },
+  ]
+
+> { tag = 'Number, value = 1+1 } | NumberOrString
+{ tag = '"Number", value | Number = 2, }
+
+> { tag = 'String, value = "hello"} | NumberOrString
+{ tag = '"String", value | String = "hello", }
+
+> { tag = 'Number, value = "hello"} | NumberOrString
+error: contract broken by the value of `value`
+[...]
+```
+
+### Immediate version of built-in contracts
+
+In the future, we plan to provide alternative versions of built-in contracts,
+like `{ foo | Number, bar | String }` or `Array {_ | Number}`, but which are
+entirely immediate. This would allow to use them in boolean combinators with the
+intuitive semantics, at the price of needing to deeply evaluate all the content
+immediately. This might be useful when converting schemas from different tools
+to Nickel blindly without having to write a lot of custom contract wrappers.
+
+As of today, those immediate versions are not yet available. It is advised to
+rely on custom wrappers for now, as explained in the previous section.
+
+## Other combinators
+
+`any_of` has been used as a leading example, but `not` has the same limitations.
+For example, `["a"] | std.contract.not (Array Number)` will fail, as
+`std.contract.not (Array Number)` is essentially the same as `std.contract.not
+(Array Dyn)`, and will reject any array.
+
+On the other hand, `all_of` doesn't suffer from the same limitations, since it has
+to apply all contracts unconditionally. One exception worth noting is function
+contracts, which are mostly delayed, and might be overstrict with all boolean
+operators including `all_of`: `all_of [Number -> Number, String -> String]` will
+reject all values, including the identity function `fun x => x`, because it's
+the same contract as `std.contract.all_of [Number, String] ->
+std.contract.all_of[Number, String]`, but `std.contract.all_of [Number,
+String]` is void.
```

Comment:
Is this really overly strict? `(id | Number -> Number) "hi"` is an error, so the function contract `Number -> Number` means "a function that only takes numbers as inputs, and returns numbers". It doesn't mean "a function that, when given a number as input, returns a number." With these semantics, `id | Number -> Number` is the identity on numbers, `id | String -> String` is the identity on strings, and `id | all_of [Number -> Number, String -> String]` is an error (or at least, a function that cannot be called)

Reply:
I added one line of justification (among other things) of why you could expect this contract to work in 78b877c23e4fdb7334459e0591f6964d98c1cb05.

Reply:
> Usually, the interpretation of a function contract T ->U is "if you give me a T, I give you an U. Otherwise, I can't promise anything"

Sure, but the interpretation in Nickel is different, right? Because `(f | T -> U) x` blames if `x` is not a `T`.

Reply:
Ok, I think I see your point now. You're saying we could do nothing if the argument isn't a `T`, which would be a strange design, but is possible.

Then the strongest argument is maybe the correspondence with types, which is an important design guideline in Nickel, that contracts derived from static types are just run-time version and should morally check the same thing, even if they might accept more values (or, put differently, they check that something is a realizer of a type). Once again, in most interpretations and implementations of union types that I know of, the union is interpreted as a set-theoretic union *on terms that can be assigned the type*. Thus `T -> U \/ V -> W` is the type of the overloaded functions that supports both `T -> U` and `V -> W`, and not `T /\ V -> U /\ W`.

Reply:
@jneem what do you think about the current version? Do you suggest I drop even more of the "doesn't work as expected" and just warn that this is the current semantics, or is it ok as it is?